pyfr/plugins/tavg.py

```diff
@@ -132,20 +159,66 @@ def _eval_acc_exprs(self, intg):
             exprs.append([npeval(v, subs) for v in self.aexprs])
 
         # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
+        return [np.array(exs) for exs in exprs]
 
-    def _eval_fun_exprs(self, intg, accex):
-        exprs = []
+    def _eval_fun_exprs(self, avars):
 
-        # Iterate over each element type our averaging region
-        for avals in accex:
-            # Prepare the substitution dictionary
-            subs = dict(zip(self.anames, avals.swapaxes(0, 1)))
+        # Prepare the substitution dictionary
+        subs = dict(zip(self.anames, avars))
+    
+        # Evaluate the function and return
+        return np.array([npeval(v, subs) for v in self.fexprs])
 
-            exprs.append([npeval(v, subs) for v in self.fexprs])
+    def _eval_fun_var(self, dev, accex):
+        dfexpr, exprs = [], []
+        eps, an = np.sqrt(self.eps), self.anames
 
-        # Stack up the expressions for each element type and return
-        return [np.dstack(exs).swapaxes(1, 2) for exs in exprs]
+        # Iterate over each element type our averaging region
+        for av in accex:
+            df = []
+
+            # Evaluate the function
+            fx = self._eval_fun_exprs(av)
+            exprs.append(fx)
+
+            for i, _ in range(len(an)):
+                # Calculate step size
+                h = eps * np.maximum(abs(av[i]), eps, where=abs(av[i])>eps, out=np.ones_like(av[i]))                
+                av[i] += h
+                
+                # Evaluate function for the step
+                fxh = self._eval_fun_exprs(av)
+                
+                # Calculate derivatives for functional averages
+                df.append((fxh - fx) / h)
+                av[i] -= h
+
+            # Stack derivatives     
+            dfexpr.append(np.array(df))
+
+        # Multiply by variance and take RMS value
+        fv = [np.linalg.norm(df * sd[:, None], axis=0) for df, sd in zip(dfexpr, dev)]
+        
+        return exprs, fv 
+
+    def _acc_avg_var(self, intg, currex):  
+        prevex, vaccex, accex = self.prevex, self.vaccex, self.accex  
+
+        # Weights for online variance and average
+        Wmp1mpn = intg.tcurr - self.prevt
+        W1mpn = intg.tcurr - self.tstart_acc
+        W1m = W1mpn - Wmp1mpn
+
+        # Iterate over element type
+        for v, a, p, c in zip(vaccex, accex, prevex, currex):
+            # Accumulate average
+            a += Wmp1mpn*(p + c)
+
+            # Accumulate variance          
+            v += Wmp1mpn*(c**2 + p**2 - 0.5*(p + c)**2) 
+            if not self.tstart_acc == self.prevt:
+                v +=  ((W1m / (2 * Wmp1mpn * W1mpn)) * (Wmp1mpn / W1m)**2
```

Comment:
Can probably clean this up by moving this constant up top in place of W1m (which is only used here).  Also, this will simplify.